eng/common/matrix/README.md

```diff
@@ -0,0 +1,242 @@
+# Azure Pipelines Matrix Generator
+
+* [Usage in a pipeline](#usage-in-a-pipeline)
+* [Matrix config file syntax](#matrix-config-file-syntax)
+   * [Fields](#fields)
+	  * [matrix](#matrix)
+	  * [include](#include)
+	  * [exclude](#exclude)
+	  * [displayNames](#displaynames)
+* [Matrix Generation behavior](#matrix-generation-behavior)
+  * [all](#all)
+  * [sparse](#sparse)
+  * [include/exclude](#includeexclude)
+  * [displayNames](#displaynames-1)
+  * [Filters](#filters)
+  * [Under the hood](#under-the-hood)
+
+
+This directory contains scripts supporting dynamic, cross-product matrix generation for azure pipeline jobs.
+It aims to replicate the [cross-product matrix functionality in github actions](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-running-with-more-than-one-version-of-nodejs),
+but also adds some additional features like sparse matrix generation, cross-product includes and excludes, and programmable matrix filters.
+
+This functionality is made possible by the ability for the azure pipelines yaml to take a [dynamic variable as an input
+for a job matrix definition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-job-configuration) (see the code sample at the bottom of the linked section).
+
+## Usage in a pipeline
+
+In order to use these scripts in a pipeline, you must provide a config file and call the matrix creation script within a powershell job.
+The job must then be marked as a dependency by a subsequent matrix job, and the matrix value pulled from a variable set by the powershell job.
+For example:
+
+```
+parameters:
+  - name: ProductMatrix
+    type: string
+    default: 'matrix.json'
+
+jobs:
+  - job: generate_matrix
```

Comment:
I suspect this should actually be a shared yml template for folks to include as well.